### PR TITLE
OCPBUGS-28981: Add exponential backoff to the container stop loop

### PR DESCRIPTION
**Pull request by kwilczynski**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Currently, when CRI-O attempts to stop a container where the process within, especially an init process (the so-called "PID 1"), is in an uninterruptible blocking state (for example, it's sleeping and waiting for a disk I/O completion, etc.), CRI-O will enter a broken state where it tries to delivery termination signals to such a process as fast as possible.

Nonetheless, a blocked process might not promptly respond to the signals delivered, causing CRI-O to enter a "busy loop" while it repeatedly tries to signal delivery. This seemingly unbound loop can render CRI-O unresponsive and result in high CPU usage while it happens.

Thus, add exponential backoff support to the container stop loop to fix the possible busy loop issue irregardless of the current state of the process to be terminated. The exponential backoff will stagger termination signals delivery for as long as the process is still running, allowing it to eventually terminate on its own volition (or crash, whichever comes first).

Related to:

- https://github.com/cri-o/cri-o/pull/7736

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

None

#### Does this PR introduce a user-facing change?

```release-note
- Add exponential backoff to the container stop loop to fix a busy loop issue when a process running with the container is in an uninterruptible blocking state and would become unresponsive to signals delivery during container termination.
- Add a new error log line to notify the user that a process (the container init) has been blocked in uninterruptible sleep for some time.```

### CODE REVIEW

File: internal/oci/runtime_oci.go
Comment (haircommander) [inline internal/oci/runtime_oci.go:873]: I think you can use the ExponentialBackoff in k8s.io/apimachinery/pkg/util/wait/ which I believe will simplify this  a bit

Comment (kwilczynski) [reply internal/oci/runtime_oci.go:873]: @haircommander, can do. 👍 

I looked at this first, as we use it throughout the code base, but then decided that having a ability to burst and then also to roll-over to eventually try shorter intervals again and burst, was an interesting idea. Simple to add too.

That said, I will convert it. So we will have progression with an increasing internal between retries up to the limit of 5 minutes, from where it will always use 5 minutes as an interval going forward.

Comment (kwilczynski) [reply internal/oci/runtime_oci.go]: Done.

File: internal/oci/runtime_oci.go
Comment (kwilczynski) [inline internal/oci/runtime_oci.go:51]: **To the reviewer:**

Hopefully, these values are somewhat acceptable. We can always raise the maximum and reset backoff limits as needed to make the loop stagger slightly more.

File: internal/oci/runtime_oci.go
Comment (saschagrunert) [inline internal/oci/runtime_oci.go:50]:
```suggestion
const (
```

And merge with the block above.

**Review by saschagrunert — COMMENTED** [review]
Is it somehow possible to write an integration test for this?

**Comment by kwilczynski** [comment]
> Is it somehow possible to write an integration test for this?

@saschagrunert, I am open to suggestions.

I use a custom kernel module to switch process states while manually testing and verifying.

File: internal/oci/runtime_oci.go
Comment (kwilczynski) [reply internal/oci/runtime_oci.go]: Done.

**Comment by saschagrunert** [comment]
> I use a custom kernel module to switch process states while manually testing and verifying.

That seems to be too much from my point of view. I'm wondering if we could create a workload triggering the backoff at least.

**Comment by andreaskaris** [comment]
> > I use a custom kernel module to switch process states while manually testing and verifying.
> 
> That seems to be too much from my point of view. I'm wondering if we could create a workload triggering the backoff at least.

No kernel module needed, it can be done with a script: https://github.com/andreaskaris/uninterruptible-sleeper/blob/master/uninterruptible.sh | https://github.com/andreaskaris/uninterruptible-sleeper/blob/master/fedora.yaml | https://chrisdown.name/2024/02/05/reliably-creating-d-state-processes-on-demand.html
But you still need to be root / privileged

**Comment by kwilczynski** [comment]
>  [...]
> Or write a dummy runtime that doesn't react to the KILL signal and test with that?

@andreaskaris, this is what I settled with too. Probably the cleanest approach.

File: internal/oci/runtime_oci.go
Comment (haircommander) [inline internal/oci/runtime_oci.go:889]: nit: personal preference to not define the variable if we don't need it. we can inline it in the BackoffUntil function below

File: internal/oci/runtime_oci.go
Comment (haircommander) [inline internal/oci/runtime_oci.go:842]: I assume you've hit this while debugging. I think it's in part incomplete because I would guess there are other pieces in cri-o we're not correctly catching SIGINT. I am fine with this for now, but maybe we should plumb all the CRI calls with this so we have better INT handling throughout?

File: internal/oci/runtime_oci.go
Comment (haircommander) [inline internal/oci/runtime_oci.go:877]: yeah I suppose the end of unix time is a bit much...

File: internal/oci/runtime_oci.go
Comment (kwilczynski) [reply internal/oci/runtime_oci.go:842]: @haircommander, yes.

We need to plumb context correctly where it should have been passed. But this would be more work and out of scope for this change.

A work cut out for a future Pull Request.

File: internal/oci/runtime_oci.go
Comment (kwilczynski) [reply internal/oci/runtime_oci.go:889]: I thought of doing it, but when done the readibility took a bit of a hit for the worse. The large block is in the middle which makes it somewhat unwildely.

Happy to change it for you. 😀

Comment (kwilczynski) [reply internal/oci/runtime_oci.go]: Done.

**Comment by kwilczynski** [comment]
> Is it somehow possible to write an integration test for this?

@saschagrunert, done.

**Review by saschagrunert — APPROVED** [review]

**Comment by kwilczynski** [comment]
/hold

Another issue has been found while working on this fix. I will update the Pull Request accordingly.

**Comment by kwilczynski** [comment]
/unhold

The other issue will be taken care of separately. I don't want to block a release of this specific fix.

**Comment by haircommander** [comment]
kata test failures are legit https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7854/pull-ci-cri-o-cri-o-main-ci-fedora-kata/1767191175355174912/artifacts/fedora-kata/cri-o-gather/artifacts/testout.txt

**Comment by kwilczynski** [comment]
> kata test failures are legit https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7854/pull-ci-cri-o-cri-o-main-ci-fedora-kata/1767191175355174912/artifacts/fedora-kata/cri-o-gather/artifacts/testout.txt

@haircommander, let me see! Thank you!

File: contrib/test/ci/vars.yml
Comment (kwilczynski) [inline contrib/test/ci/vars.yml:132]: **To the reviewer:**

The Runtime VM (Kata) implements its own `StopContainer()` with custom logic, log lines, etc. Hence, the OCI test does not apply to it.

**Comment by kwilczynski** [comment]
/hold

Holding for a moment...

File: internal/oci/runtime_oci.go
Comment (kwilczynski) [inline internal/oci/runtime_oci.go:846]: **To the reviewer:**

I decided to allow for passing a BackoffManager implementation so that we can tweak the one used in the unit tests. Otherwise, it would increase the total run time of the test (nearly double), which is not ideal for a unit test.

**Comment by kwilczynski** [comment]
@cri-o/cri-o-maintainers, I think this could now go in as is.

**Review by saschagrunert — APPROVED** [review]

**Comment by mrunalp** [comment]
I don't see the code to detect and log that the process in in D state. I think that's important to add so we can easily classify such issues by looking at the logs.

**Comment by kwilczynski** [comment]
> I don't see the code to detect and log that the process in in D state. I think that's important to add so we can easily classify such issues by looking at the logs.

@mrunalp, the previous implementation wasn't fit for use and was not picked up.

I would like to get this quite urgent bug fix on the way, which addresses the issue of a busy loop in the stop code. We can revisit adding improved reporting on the state of the processes within the container as a follow-up. Would this be acceptable?

**Comment by kwilczynski** [comment]
/hold

I found another issue while doing some manual testing... Putting a hold for now.

**Comment by kwilczynski** [comment]
> > I don't see the code to detect and log that the process in in D state. I think that's important to add so we can easily classify such issues by looking at the logs.
> 
> @mrunalp, the previous implementation wasn't fit for use and was not picked up.
> 
> I would like to get this quite urgent bug fix on the way, which addresses the issue of a busy loop in the stop code. We can revisit adding improved reporting on the state of the processes within the container as a follow-up. Would this be acceptable?

Done.

I've added an error log that will be shown when a process is blocked for some time. The log will be repeated periodically as a constant reminder.

File: internal/oci/runtime_oci.go
Comment (andreaskaris) [inline internal/oci/runtime_oci.go:892]: I'm not sure that this happens *only* on uninterruptible sleep (it's likely only on uninterruptible sleep, but in another customer case under investigation, I could not find evidence that the process in question was in uninterruptible sleep). Anyway, this state is for the runtime's init process, right?
Nit: "Detected process (%d) blocked in uninterruptible sleep for more than %d seconds for container: %s" would IMO make it clearer, especially given that you test for "D"

Comment (kwilczynski) [reply internal/oci/runtime_oci.go:892]: @andreaskaris, most of the time, we only track the init PID within the container within CRI-O, so if the process is blocked or waiting either for one of its children or a thread (albeit, this case is slightly different), then you could also see things blocked despite the primary process within the container sleeping with "S" state, etc.

We can add a message that spells out a list of children and threads of the init process with their state, which could help a little more. This is something that I ran through @mrunalp already, but it's also something he wants to include as a later, perhaps follow-up, change.

Comment (kwilczynski) [reply internal/oci/runtime_oci.go:892]: [...]
> With the test for "D" only: "Detected process (%d) blocked **in uninterruptible sleep** for more than %d seconds for container: %s" would IMO make it clearer, especially given that you test for "D"

Done.

**Comment by kwilczynski** [comment]
@cri-o/cri-o-maintainers @andreaskaris, please have a look. We need to get this out as soon as possible.

Thank you!

File: internal/oci/runtime_oci.go
Comment (andreaskaris) [reply internal/oci/runtime_oci.go:892]: thanks!

File: internal/oci/container_linux.go
Comment (andreaskaris) [inline internal/oci/container_linux.go:90]: Not sure if this is a nit or a minuscule bug ..

Not that it would matter because the stat file is 52 fields long (50 when factoring out pid and name), but startTimeFieldLocation 22 is too high, given that you're looking into fields[19].

By explicitly naming the fields, the code would be more readable, and the offset / length error could be avoided:
```
const startTimeFieldLocation = 19
const stateFieldLocation = 0
(...)
if len(fields) < startTimeFieldLocation + 1 {
(...)
return string(fields[procFieldLocation]), string(fields[startTimeFieldLocation]), nil
```

**Comment by andreaskaris** [comment]
Feel free to disregard my nits. Most of the code is too complex for me, so I defer to the maintainers for review. I do like the error message now and the general approach with the backoff looks o.k. to me

File: internal/oci/container_linux.go
Comment (kwilczynski) [reply internal/oci/container_linux.go]: Surprisingly, the code originally looked like that, but then I decided to eliminate the variables to make it more concise.

Comment (kwilczynski) [reply internal/oci/container_linux.go]: Done. Thank you!

**Comment by kwilczynski** [comment]
> Feel free to disregard my nits. Most of the code is too complex for me, so I defer to the maintainers for review. I do like the error message now and the general approach with the backoff looks o.k. to me

@andreaskaris, no worries! I appreciate every review! :muscle:

**Comment by kwilczynski** [comment]
@cri-o/cri-o-maintainers, I believe this should be good to go.

**Comment by kwilczynski** [comment]
@haircommander, added the `ProcessState()` function as requested, in lieu of modifying `Living()`.

**Comment by kwilczynski** [comment]
OK. Requires manual backport.

**Pull request merged**
